File: 프로그래머스/3/43162. 네트워크/네트워크.cpp
#include <string>
#include <vector>
using namespace std;

vector<int> parent(200, -1);

int find_parent(int c1)
{
    while(parent[c1] != - 1)
		c1 = parent[c1];
	return c1;
}

void union_com(int c1, int c2)
{
    c1 = find_parent(c1);
    c2= find_parent(c2);

    if (c1 > c2)
        parent[c1] = c2;
    else if (c1 < c2)
        parent[c2] = c1;
	else
		return;
}

int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    
    for(int i = 0; i < n; i++)
    {
        for(int j = 0; j < n; j++)
        {
            if(computers[i][j] == 1)
                union_com(i, j);
        }
    }
    
    for (int k = 0; k < n; k++)
    {
        if(parent[k] == -1)
            answer++;
    }
            
    return answer;
}

File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.Queue;
import java.util.LinkedList;
import java.util.Arrays;


class Solution {
    boolean[] visited;
    int network = 0;
    
    void BFS(int[][] computers, int n, int idx) {
        Queue<Integer> que = new LinkedList();
        que.add(idx);
        visited[idx] = true;
        
        while (!que.isEmpty())
        {
            int front = que.element();
            que.remove();
            
            for (int i = 0; i < n; i++) {
                if (i != front && computers[front][i] == 1 && visited[i] == false) {
                    visited[i] = true;
                    que.add(i);
                }
            }
        }    
    }
    
    public int solution(int n, int[][] computers) {
        visited = new boolean[n];
        Arrays.fill(visited, false);
      
        for (int i = 0; i < n; i++) {
            if (visited[i] == false) {
                BFS(computers, n, i);
                network += 1;
            }
        }
        
        return network;
    }
}